src/TBuffer.cpp

```diff
@@ -4039,10 +4039,37 @@ bool TBuffer::processGBSequence(const std::string& bufferData, const bool isFrom
                 dataIdentity = "User Defined (PU) Area 3";
 #endif
 
+                // clang-format off
+            } else if (  (static_cast<quint8>(bufferData.at(pos    )) >= 0x90) && (static_cast<quint8>(bufferData.at(pos    )) <= 0xE3)
+                      && (static_cast<quint8>(bufferData.at(pos + 1)) >= 0x30) && (static_cast<quint8>(bufferData.at(pos + 1)) <= 0x39) ) {
+// clang-format on
+// First two bytes of a 4-byte GB18030 sequence - for a non-BMP mapped Unicode
+// codepoint if byte 3 is within 0x81-00xFE and byte 4 is within 0x30-0x39
+                isValid = false;
+                isToUseReplacementMark = true;
+#if defined(DEBUG_GB_PROCESSING)
+                qDebug().nospace() << "TBuffer::processGBSequence(...) 2-byte sequence as "
+                                      "GB2312/GBK rejected because it is seems to be the "
+                                      "first pair of a 4-byte GB18030 non-BMP Unicode sequence!";
+#endif
+                // clang-format off
+            } else if (  (static_cast<quint8>(bufferData.at(pos    )) >= 0xFD) && (static_cast<quint8>(bufferData.at(pos    )) <= 0xFE)
+                      && (static_cast<quint8>(bufferData.at(pos + 1)) >= 0x30) && (static_cast<quint8>(bufferData.at(pos + 1)) <= 0x39) ) {
+// clang-format on
+// First two bytes of a 4-byte GB18030 sequence - for a non-BMP mapped Unicode
+// codepoint if byte 3 is within 0x81-00xFE and byte 4 is within 0x30-0x39
+                isValid = false;
+                isToUseReplacementMark = true;
+#if defined(DEBUG_GB_PROCESSING)
+                qDebug().nospace() << "TBuffer::processGBSequence(...) 2-byte sequence as "
+                                      "GB2312/GBK rejected because it is seems to be the "
+                                      "first pair of a 4-byte GB18030 Private Use sequence!";
+#endif
             } else {
-                // Outside expected range
+                // Outside expected ranges
 
                 isValid = false;
+                isToUseReplacementMark = true;
```

Comment:
This is the core of the fix - but I had put two more range checks to decode what I was seeing above right above here before I spotted that it was missing: by not inserting a `QChar` (the replacement mark) but still inserting a `TChar` to format it I was provoking the paint code to crash.  However it ***was*** triggering a `qWarning()` at the point (around line 2180 in this file) where I check that there is the same number of them for each line added into the `TBuffer` instance so I had some idea what was wrong... :slightly_smiling_face: